src/js/components/history/HistoryItem.tsx

```diff
@@ -0,0 +1,192 @@
+    const dropped = tupleid(sel) && tupleid(cell);
+
+    try {
+      const lyraId = +sel.mark.role.split('lyra_')[1]; // id of thing that was dropped onto
+
+      if (dropped) {
+        const channel = channelName(cell.key);
+        let fieldName, dsId;
+        if (channel === 'x' || channel === 'y' || channel === 'color' || channel === 'size') {
+          // set scale
+          let channelScaleIds = this.props.history.getIn(["guides"])
+            .map((g) => {
+              return channel == 'x' || channel == 'y' ? g.scale : channel == 'size' ? g[channel] : g.fill;
+            })
+            .filter((scaleId) => {
+              let scaleName = scaleId ? this.props.history.getIn(["scales"]).get(scaleId).name : null;
+              return scaleName === channel;
+            });
+
+          fieldName = channelScaleIds.map((scaleId) => {
+            let scaleRecord = this.props.history.getIn(["scales"]).get(scaleId);
+            return scaleRecord.get('_domain').length > 0 ?  scaleRecord.get('_domain')[0].field : null;
+          }).first();
+
+          dsId = channelScaleIds.map((scaleId) => {
+            let scaleRecord = this.props.history.getIn(["scales"]).get(scaleId);
+            return scaleRecord.get('_domain').length > 0 ?  scaleRecord.get('_domain')[0].data : null;
+          }).first();
+
+
+        }
+
+        const bindField = this.props.history.getIn(["datasets", String(dsId), "_schema", fieldName]).toJS();
+        vega.extend(bindField, opts); // Aggregate or Bin passed in opts.
+        props.bindChannel(dsId, bindField, lyraId, cell.key);
+      } else {
+        // update the whole symbol/mark: shape, size
+        // the history has the signals
+        let relevantProps = ["size", "shape", "fill", "fillOpacity", "stroke",  "strokeWidth"];
+        let historyMark = this.props.history.getIn(["marks", String(lyraId)]); // assume mark is the same as before. wanna revert to old settings of old mark
+
+        batchGroupBy.start();
+        relevantProps.forEach((prop) => {
+          let historyProp = historyMark.getIn(["encode", "update", prop]);
+          if (historyProp.signal) {
+            let historySigVal = this.props.history.getIn(["signals", historyProp.signal]).value;
+            this.props.setSignal(historySigVal, historyProp.signal); // update Store
+            sg.set(historyProp.signal, historySigVal, false); // update Vega
+          }
+        });
+        batchGroupBy.end();
+      }
+    } catch (e) {
+      console.error('Unable to bind primitive');
+      console.error(e);
+    }
+
+    this.props.stopDragging();
+    sg.set(MODE, 'handles');
+    sg.set(CELL, {});
+
+    ctrl.update(); // Apply changes
+
+  }
+
+  private historyToSpec(preview: HistoryRecord): Spec {
+    let historySpec = ctrl.export(false, this.props.history);
+    if (historySpec.marks) {
+      historySpec = cleanSpecForPreview(historySpec, this.width, this.height, null, parseInt(this.props.id), true);
```

Comment:
see the comment near the top of the file about whether `id` should be typed as a `number`